src/Data/Primitive/Contiguous/Slice.hs

```diff
@@ -0,0 +1,160 @@
+  freeze MutableSlice{array,offset,length} off' len'
+    = fromArray <$> freeze array (offset + off') (min length len')
+  thaw Slice{array,offset,length} off' len' = do
+    fromMutableArray =<< thaw array (offset + off') (min length len')
+  copy dst dstOff src srcOff len = do
+    let MutableSlice{array=arrD,offset=offD} = dst
+        Slice{array=arrS,offset=offS} = src
+    copy arrD (dstOff + offD) arrS (srcOff + offS) len
+  copyMutable dst dstOff src srcOff len = do
+    let MutableSlice{array=arrD,offset=offD} = dst
+        MutableSlice{array=arrS,offset=offS} = src
+    copyMutable arrD (dstOff + offD) arrS (srcOff + offS) len
+  clone Slice{array,offset,length} off len'
+    = Slice{array,offset=offset+off,length=min length len'}
+  cloneMutable MutableSlice{array,offset,length} off len' = do
+    fromMutableArray =<< cloneMutable array (offset + off) (min length len')
+  equals Slice{array=a,offset=oA,length} Slice{array=b,offset=oB,length=lenB}
+    = length == lenB && loop 0 oA oB
+    where
+    loop !i !iA !iB =
+      if i == length then True
+      else index a iA == index b iB && loop (i+1) (iA+1) (iB+1)
+  equalsMutable MutableSlice{array=arrA,offset=offA,length=lenA}
+                MutableSlice{array=arrB,offset=offB,length=lenB}
+    =  arrA `equalsMutable` arrB
+    && offA == offB
+    && lenA == lenB
+  -- unlift
+  -- lift
+  singleton x = fromArray $ singleton x
+  doubleton x y = fromArray $ doubleton x y
+  tripleton x y z = fromArray $ tripleton x y z
+  quadrupleton x y z w = fromArray $ quadrupleton x y z w
+  rnf Slice{array} = rnf array `seq` ()
+  run = runST
+  {-# inline empty #-}
+  {-# inline null #-}
+  {-# inline new #-}
+  {-# inline replicateMutable #-}
+  {-# inline index #-}
+  {-# inline index# #-}
+  {-# inline indexM #-}
+  {-# inline read #-}
+  {-# inline write #-}
+  {-# inline resize #-}
+  {-# inline size #-}
+  {-# inline sizeMutable #-}
+  {-# inline unsafeFreeze #-}
+  {-# inline freeze #-}
+  {-# inline thaw #-}
+  {-# inline copy #-}
+  {-# inline copyMutable #-}
+  {-# inline clone #-}
+  {-# inline cloneMutable #-}
+  {-# inline equals #-}
+  {-# inline equalsMutable #-}
+  -- {-# inline unlift #-}
+  -- {-# inline lift #-}
+  {-# inline singleton #-}
+  {-# inline doubleton #-}
+  {-# inline tripleton #-}
+  {-# inline quadrupleton #-}
+  {-# inline rnf #-}
+  {-# inline run #-}
+
+-- TODO I think I'll want to add slice and sliceMutable to Contiguous, which
+-- means moving this module into the big one, or making a
+-- Data.Primitive.Contiguous.Class module
+
+-- TODO freezeAll, thawAll, copyAll, copyMutableAll, cloneAll, cloneMutableAll
+
+-- This interface is definitely strange for slices.
+-- It seems like if we're going to combine array slicing and coniguous, the
+-- whole class needs to be altered.
+
+-- TODO
+-- Contiguous ByteArray where Element = (~ Word8)
+
+I think the thing that makes the most sense is to create a "competing" library to contiguous.
```

Comment:
It looks like this was accidentally copy-pasted from a chat.

Reply:
ah yes, a preliminary file that needed deleting